src/v/datalake/translation/partition_translator.h

```diff
@@ -0,0 +1,166 @@
+/*
+ * Copyright 2024 Redpanda Data, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/redpanda-data/redpanda/blob/master/licenses/rcl.md
+ */
+
+#pragma once
+
+#include "base/outcome.h"
+#include "base/seastarx.h"
+#include "cluster/fwd.h"
+#include "datalake/errors.h"
+#include "datalake/fwd.h"
+#include "features/fwd.h"
+#include "model/record_batch_reader.h"
+#include "random/simple_time_jitter.h"
+#include "ssx/semaphore.h"
+#include "utils/prefix_logger.h"
+
+#include <seastar/core/abort_source.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/semaphore.hh>
+#include <seastar/core/shared_ptr.hh>
+
+namespace kafka {
+class read_committed_reader;
+}
+
+namespace datalake::translation {
+
+/**
+ * A partition translator is responsible for translating data from a given
+ * partition periodically. A single instance of translator should be active for
+ * a given partition and is attached to the leader replica.
+ *
+ * The translator periodically wakes up, checks if there is new data that is
+ * pending translation and creates a reader for the data range and hands it off
+ * to low level iceberg translator that converts the data.
+ *
+ * The translator works in tandem with the datalake coordinator responsible for
+ * this ntp. A successful local iceberg translation of a data range is followed
+ * by that state checkpointing with the coordinator. A coordinator tracks the
+ * latest translated offset for a given ntp (along with other metadata) which is
+ * also then synchronized with the local translation stm that enforces
+ * max_collectible_offset across all replicas. The translator tries to work
+ * with the stm state as much as possible (to avoid round trips to the
+ * coordinator)
+ *
+ * It is possible that the translation stm can get out of sync with what is
+ * on the coordinator, so it is reconciled as needed. It is ok to work on a
+ * stale stm state as the eventual checkpoint request with the coordinator
+ * fenced based on the translator term.
+ *
+ * The logic for this translator is like this..
+ *
+ * while (!aborted && !term_changed):
+ *    sleep(interval)
+ *    reconcile_with_coordinator_if_needed()
+ *    md = translate_newly_arrived_data_since_last_checkpoint()
+ *    checkpoint_with_coordinator(md)
+ *    sync_stm_with_coordinator(md)
+ */
+
+class partition_translator {
```

Comment:
How much effort is it to get a unit test for this? At least laydown the foundations for better tests later (I guess unlikely, but is nice to hope!).